Add Storybook documentation and remove deprecated knobs

sunyatasattva opened the pull request:
Now that the Storybook [migration of existing stories](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5248) is complete, we can remove the knobs dependencies as they have been fully replaced by controls.

This PR is mostly just removing unused things, and adding a `README` file and VSCode snippets for smoother Storybook development.

However, while removing the knobs, I have noticed another story that hadn't been migrated because it was not noticed due to an error which didn't make it appear in the list of available stories. For now, I quickly migrated them to the new controls, but didn't really polish them: they all render but they are not as refined as the other ones in the epic, so I have added `todo` comments to make sure we follow-up and make it clear about their status. 

### Testing

### Manual Testing

Not really much testing to do for this PR. Perhaps just check the docs make some sense.

sunyatasattva committed:
Remove Storybook knobs

They are now fully replaced by controls

sunyatasattva committed:
Update Storybook to 6.4.9

sunyatasattva committed:
Add Storybook documentation and VSCode snippets

sunyatasattva committed:
Add VSCode snippets

sunyatasattva committed:
Convert stories from checkout packages and include them in Storybook

sunyatasattva committed:
Change todo comment to avoid duplicates

sunyatasattva committed:
Remove knobs

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+# Working with Storybook
+
+This document is meant to make contributing to our Storybook a bit easier by giving some tips, pointing out a few gotchas and make the editing experience smoother and reduce the friction to adding a new story.
```
```suggestion
This document is meant to make contributing to our Storybook a bit easier by giving some tips, pointing out a few gotchas, making the editing experience smoother, and reducing the friction to adding a new story.
```

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+
+First of all, by using TypeScript and exporting your component's props, Storybook will automatically show the documentation associated with your props and also generate the controls which are more appropriate for each property.
+
+Also, if you follow the naming convention of starting your event handlers with `on` (e.g. `onChange`, `onRemove` etc.), Storybook will automatically make those props uncontrollable and generate events from them (this convention can be changed in `preview.js`).
```
```suggestion
Also, if you follow the naming convention of starting your event handlers with `on` (e.g. `onChange`, `onRemove`, etc.), Storybook will automatically make those props uncontrollable and generate events from them (this convention can be changed in `preview.js`).
```

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+  title: string;
+
+  /**
+   * You should basically always pass your component here.
```
```suggestion
   * You should always pass your component here.
```

dinhtungdu commented on storybook/README.md:
````diff
@@ -0,0 +1,322 @@
+);
+```
+
+In this way you are rendering the component in the viewport and binding the component properties to the Storybook controls.
````
```suggestion
In this way, you are rendering the component in the viewport and binding the component properties to the Storybook controls.
```

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+
+### What should constitute a story?
+
+Stories should show the component in its different shapes.
```
```suggestion
Stories should show the component in their different shapes.
```

sunyatasattva replied:
Here “its” refers to the component, not to the “stories”, so “its” should be more correct :|

dinhtungdu replied:
Got it! : )

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+
+While stories for closely related components could technically be living within the same file, we advise that you keep one file per component, as this works best with the automatic organization of the components within storybook and using [`default` exports in the Story](https://storybook.js.org/docs/react/api/csf#default-export).
+
+An exception to this, might be when there is only one mainly used component and the other defined ones are just used internally.
```
```suggestion
An exception to this might be when there is only one mainly used component and the other defined ones are just used internally.
```

sunyatasattva replied:
I feel this comma sounds a bit awkward in the middle of the sentence. Perhaps, if the sentence would have been `As an exception to this, […there is that]`, I would have seen it as more appropriate… 🤔

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+
+Sometimes the inferred text input is not good enough, and you need something more complex. If those controls happen to be shared among many components, please write them in `storybook/custom-controls`.
+
+An example there is the `currency` control. Since many of our components expect `Currency` objects in their props, we can make sure we give a few example Storybook users can play with, without having to manually create the object themselves.
```
```suggestion
An example there is the `currency` control. Since many of our components expect `Currency` objects in their props, we can make sure we give a few examples Storybook users can play with, without having to manually create the object themselves.
```

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+
+### Controlled components
+
+Your component is not managing its own state, and expects it to be passed as a prop, but you want to create a self-contained story. You can then edit your main `Template` function to manage the state, for example through hooks.
```
```suggestion
Your component is not managing its own state and expects it to be passed as a prop, but you want to create a self-contained story. You can then edit your main `Template` function to manage the state, for example through hooks.
```

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+
+### Simulating interactions
+
+First of all, note that all the props starting with `on` as described above trigger “actions” in Storybook, which is basically just a way to see how data gets passed to an handler via the _Actions_ tab.
```
```suggestion
First of all, note that all the props starting with `on` as described above trigger “actions” in Storybook, which is basically just a way to see how data gets passed to a handler via the _Actions_ tab.
```

dinhtungdu commented on storybook/README.md:
```diff
@@ -0,0 +1,322 @@
+
+If you are using [VSCode](https://code.visualstudio.com/), this repo includes some helpful snippets that will become available inside `.ts` and `.tsx` files. You can find them inside `.vscode/storybook.code-snippets`.
+
+The `sbs` (“Storybook story”) will scaffold the entire code in the [section above](#scaffold-tour). If you have respected the naming conventions, it will also property import your component and the properties with the correct names, saving you a bunch of time.
```
```suggestion
The `sbs` (“Storybook story”) will scaffold the entire code in the [section above](#scaffold-tour). If you have respected the naming conventions, it will also properly import your component and the properties with the correct names, saving you a bunch of time.
```

dinhtungdu reviewed (commented):
@sunyatasattva Great work on the documentation as well as the migration 👏🏽 . I added some suggestions on Grammar and wording.

I check the storybook and got an error:

```
Unexpected error while loading ./base/components/cart-checkout/totals/shipping/stories/index.tsx: Cannot read properties of undefined (reading 'id')
```

By the way, I notice some duplicated todos added in this PR, we should remove those duplications or note explicitly the exact component for each todo.

sunyatasattva committed:
Update storybook/README.md

Co-authored-by: Tung Du <dinhtungdu@gmail.com>

sunyatasattva committed:
Update storybook/README.md

Co-authored-by: Tung Du <dinhtungdu@gmail.com>

sunyatasattva committed:
Update storybook/README.md

Co-authored-by: Tung Du <dinhtungdu@gmail.com>

sunyatasattva committed:
Update storybook/README.md

Co-authored-by: Tung Du <dinhtungdu@gmail.com>

sunyatasattva committed:
Update storybook/README.md

Co-authored-by: Tung Du <dinhtungdu@gmail.com>

sunyatasattva committed:
Apply suggestions from code review

Co-authored-by: Tung Du <dinhtungdu@gmail.com>

sunyatasattva committed:
Remove totals shipping story

sunyatasattva commented:
> @sunyatasattva Great work on the documentation as well as the migration 👏🏽 . I added some suggestions on Grammar and wording.

Hey @dinhtungdu, thanks a lot for the proof-reading. I accepted most of your suggestions, but added comments in two of them (I think) on why I am hesitant to include them.  Let me know what you think.

> I check the storybook and got an error:
> 
> ```
> Unexpected error while loading ./base/components/cart-checkout/totals/shipping/stories/index.tsx: Cannot read properties of undefined (reading 'id')
> ```
> 

I decided to remove that story altogether for now. That component is quite tricky as it accesses the data store via a custom hook, so it requires a bunch of extra steps for the stories. The scope of this PR was not really to add stories, but just to deprecate knobs and add docs; I had fixed the low-effort things just to pass linting pretty much.

So, that will be for another PR.

> By the way, I notice some duplicated todos added in this PR, we should remove those duplications or note explicitly the exact component for each todo.

I actually added duplicates explicitly. As far as I understood the GitHub automation, if the title is the same, no duplicate issue will be created. So I made sure to word the comment in the same way to create one issue, but to have the comment in multiple places so it's easy to pick up when someone decides to fix those stories.

dinhtungdu reviewed (approved):
> I actually added duplicates explicitly. As far as I understood the GitHub automation, if the title is the same, no duplicate issue will be created.

That's what I worried about. Thanks for the clarification, this is LGTM now! :shipit:

sunyatasattva committed:
Update package-lock
